proof/access-control/AARCH64/ArchAccess.thy

```diff
@@ -0,0 +1,363 @@
+(*
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+theory ArchAccess
+imports Types
+begin
+
+context Arch begin global_naming AARCH64
+
+subsection \<open>Arch-specific transformation of caps into authorities\<close>
+
+definition vspace_cap_rights_to_auth :: "cap_rights \<Rightarrow> auth set" where
+  "vspace_cap_rights_to_auth r \<equiv>
+     (if AllowWrite \<in> r then {Write} else {})
+   \<union> (if AllowRead \<in> r then {Read} else {})"
+
+definition arch_cap_auth_conferred where
+  "arch_cap_auth_conferred arch_cap \<equiv>
+     (if is_FrameCap arch_cap then vspace_cap_rights_to_auth (acap_rights arch_cap) else {Control})"
+
+subsection \<open>Generating a policy from the current ASID distribution\<close>
+
+definition pte_ref2 where
+  "pte_ref2 level pte \<equiv> case pte of
+     PagePTE paddr _ _ rights
+       \<Rightarrow> Some (ptrFromPAddr paddr,
+                pt_bits_left level,
+                vspace_cap_rights_to_auth rights)
+   | PageTablePTE ppn
+       \<Rightarrow> Some (ptrFromPAddr (paddr_from_ppn ppn), 0, {Control})
+   | _ \<Rightarrow> None"
+
+definition vs_refs_aux :: "vm_level \<Rightarrow> arch_kernel_obj \<Rightarrow> (obj_ref \<times> obj_ref \<times> aa_type \<times> auth) set"
+  where
+  "vs_refs_aux level \<equiv> \<lambda>ko. case ko of
+     ASIDPool pool \<Rightarrow> (\<lambda>(r,p). (p, ucast r, AASIDPool, Control)) ` graph_of (option_map ap_vspace o pool)
+   | PageTable pt \<Rightarrow> (case pt of
+       VSRootPT pt \<Rightarrow> \<Union>(r,(p, sz, auth)) \<in> graph_of (pte_ref2 level o pt).
+                      (\<lambda>(p, a). (p, ucast r, APageTable VSRootPT_T, a)) ` (ptr_range p sz \<times> auth)
+     | NormalPT pt \<Rightarrow> \<Union>(r,(p, sz, auth)) \<in> graph_of (pte_ref2 level o pt).
```

Comment:
the `|` should line up

Reply:
Here I'm aligning `VSRootPT` and `NormalPT`. The `of` should be moved to make this clearer. I also have an alternative definition for `vs_refs_aux` in `ArchArch_AC` that avoids pt splitting. It's currently tagged with a `FIXME AARCH64` since it needs more work to be integrated, but I've proved its equivalence to this definition.